### PR TITLE
Migrate to DI

### LINKED ISSUE #30: Migrate fusion-plugin-universal-events to DI

### PR DESCRIPTION
**Pull request by ganemone**
Fixes #30

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# fusion-plugin-universal-events
+  # fusion-plugin-universal-events
```
Comment (lhorie) [inline README.md]: remove leading spaces

File: README.md [hunk 2]
```diff
@@ -182,4 +169,14 @@ Sets the frequency at which data is flushed to the server. Resets the interval t
...
-Stops the interval timer, clears the data queue and prevents any further data from being flushed to the server.
+Stops the interval timer, clears the data queue and prevents any further data from being flushed to the server. Useful for testing
+
+#### `events.from`
+
+```js
+const scoped = events.from(ctx);
+```
+
+Returns a scoped version of the events api.
```
Comment (lhorie) [inline README.md]: Explain why one would want to use this over the global instance. Document API differences between a global emitter and a scoped one.

File: README.md [hunk 3]
```diff
@@ -182,4 +169,14 @@ Sets the frequency at which data is flushed to the server. Resets the interval t
...
+
+- `ctx: FusionContext` - Required.
```
Comment (lhorie) [inline README.md]: Should link to docs about FusionContext

File: src/server.js [hunk 4]
```diff
@@ -21,81 +21,86 @@
...
 import Emitter from './emitter';
 
+export class GlobalEmitter extends Emitter {
+  constructor() {
+    super();
+    this.from = memoize(ctx => {
+      return new ScopedEmitter(ctx, this);
+    });
+  }
+  emit(type, payload, ctx) {
```
Comment (lhorie) [inline src/server.js]: Under what use cases would one use the 3rd arg here, and under what use cases would one use emitter.from? Is there a strong enough motivation to have both be part of the API?

File: README.md [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# fusion-plugin-universal-events
+  # fusion-plugin-universal-events
 
 [![Build status](https://badge.buildkite.com/de4e30ddb9d019f5a8e3a2519bc0a5cccab25247809cd10c99.svg?branch=master)](https://buildkite.com/uberopensource/fusion-plugin-universal-events)
```
Comment (lhorie) [inline README.md]: There should be a section talking about why we didn't use an existing event emitter library and how the API and semantics of our library differs from other event emitters.

If this library is not a suitable implementation for a hypothetical `EventEmitterToken` interface, we should also mention that.

File: src/server.js (hunk 4 above)
Comment (ganemone) [reply src/server.js]: This is purely implementation detail, and not exposed

File: README.md (hunk 2 above)
Comment (ganemone) [reply README.md]: That should be explained in the "Accessing `ctx` section

File: README.md (hunk 3 above)
Comment (ganemone) [reply README.md]: How are we planning on doing those links when we have to handle both open source and closed source documentation?

**Comment by ganemone** [comment]
!merge

**Pull request merged**